src/beatree/leaf/free_list.rs

```diff
@@ -0,0 +1,293 @@
+            free_list_portions.push((free_list_pn, free_list));
+            free_list_pn = prev;
+        }
+
+        FreeList {
+            head: free_list_portions.pop(),
+            portions: free_list_portions,
+            released_portions: vec![],
+        }
+    }
+
+    pub fn head_pn(&self) -> Option<PageNumber> {
+        self.head.as_ref().map(|(head_pn, _)| head_pn).copied()
+    }
+
+    pub fn pop(&mut self) -> Option<PageNumber> {
+        let Some((head_pn, head)) = &mut self.head else {
+            // If there is no available head,
+            // then it means that there are no more elements in the free list
+            return None;
+        };
+
+        let leaf_pn = head.pop().unwrap();
+
+        // replace head if we just emptied one
+        if head.is_empty() {
+            self.released_portions.push(*head_pn);
+            self.head = self.portions.pop();
+        }
+
+        Some(leaf_pn)
+    }
+
+    fn push(
+        &mut self,
+        pn: PageNumber,
+        next_head_pns: &mut impl Iterator<Item = PageNumber>,
+        force_new_head: bool,
+    ) -> Option<(PageNumber, FreeListPage)> {
+        let mut encoded_head = None;
+
+        // create new_head if required
+        match &mut self.head {
+            None => {
+                self.head = Some((next_head_pns.next().unwrap(), vec![]));
+            }
+            Some((head_pn, head)) if force_new_head || head.len() == MAX_PNS_PER_FREE_PAGE => {
+                let prev = self
+                    .portions
+                    .last()
+                    .map(|(pn, _)| *pn)
+                    .unwrap_or(PageNumber(0));
+
+                encoded_head = Some((*head_pn, encode_free_list_page(prev, head)));
+
+                self.portions.push((*head_pn, std::mem::take(head)));
+                *head_pn = next_head_pns.next().unwrap();
+            }
+            _ => (),
+        };
+
+        // extract head safely
+        let head = self.head.as_mut().map(|(_, h)| h).unwrap();
+        head.push(pn);
+
+        encoded_head
+    }
+
+    /// Appends PageNumbers to the list of free pages returning all the pages
+    /// that need to be written to storage along with their relative PageNumbers
+    ///
+    /// When a new portion is required, the page number is taken from the previous head of the list.
+    ///
+    /// There will be some fragmentation in the old head of the free list,
+    /// for example:
+    ///   * Free List: 1, 7, 9, 2, X (max 5 PageNumbers per portion)
+    ///   * Appending: 10, 18
+    ///   * Result: 1, 7, 9, 10, X | 18
+    ///
+    /// 18 is allocated into a new portion with page number 2,
+    /// and the previous head will be rewritten with a little fragmentation inside (4 bytes).
+    ///
+    /// If the list is empty, page numbers are taken from the nonce.
+    ///
+    /// Returns two vectors containing the pages to be written into the store
+    pub fn encode(
```

Comment:
Not sure about the name: `encode` sounds like it should be something idempotent or side-effect free.

Reply:
fair, it is not idempotent

I will find a better name!